packages/router/utils.ts

```diff
@@ -468,25 +469,35 @@ function explodeOptionalSegments(path: string): string[] {
   if (rest.length === 0) {
     // Intepret empty string as omitting an optional segment
     // `["one", "", "three"]` corresponds to omitting `:two` from `/one/:two?/three` -> `/one/three`
-    return isOptional ? ["", required] : [required];
+    return isOptional ? [required, ""] : [required];
   }
 
   let restExploded = explodeOptionalSegments(rest.join("/"));
-  return restExploded
-    .flatMap((subpath) => {
-      // /one + / + :two/three -> /one/:two/three
-      let requiredExploded =
-        subpath === "" ? required : required + "/" + subpath;
-      // For optional segments, return the exploded path _without_ current segment first (`subpath`)
-      // and exploded path _with_ current segment later (`subpath`)
-      // This ensures that exploded paths are emitted in priority order
-      // `/one/three/:four` will come before `/one/three/:five`
-      return isOptional ? [subpath, requiredExploded] : [requiredExploded];
-    })
-    .map((exploded) => {
-      // for absolute paths, ensure `/` instead of empty segment
-      return path.startsWith("/") && exploded === "" ? "/" : exploded;
-    });
```

Comment:
The flatmap was the weasel here since it caused he versions where this path was included to be interspersed with all child combinations when we want them all at the front.

```
// Assume
[a, b] // => [required, optional]
[c, d] // => [required, optional]
[e, f] // => [required, optional]

// We were getting this:
[a,b,c,d,e,f]

// But we wanted this:
[a,c,e,b,d,f]
```